### ISSUE TITLE
NullReferenceException / "Object reference not set to an instance of an object" when calling RejectAsync

### ISSUE THREAD

**Issue by mfalkvidd** [Azure/azure-iot-sdk-csharp#1036]
- **OS, version, SKU and CPU architecture used:**
Azure Cloud Service

- **Application's .NET Target Framework :**
4.6.1

- **Device:**
No physical device, this is an Azure Cloud Service

- **SDK version used:**
Microsoft.AspNet.WebApi.Client 5.2.7
Microsoft.AspNet.WebApi.Core 5.2.7
Microsoft.Azure.Amqp 2.4.3
Microsoft.Azure.Devices 1.18.1
Microsoft.Azure.Devices.Client 1.21.0
Microsoft.Azure.Devices.ProtocolGateway.Core 2.0.1
Microsoft.Azure.Devices.ProtocolGateway.IotHubClient 2.0.1
Microsoft.Azure.Devices.Shared 1.16.0
Microsoft.Azure.KeyVault.Core 3.0.4
Microsoft.CSharp 4.5.0
Microsoft.Data.Edm 5.8.4
Microsoft.Data.OData 5.8.4
Microsoft.Data.Services.Client 5.8.4
Microsoft.Extensions.Caching.Abstractions 2.2.0
Microsoft.Extensions.Configuration 2.2.0
Microsoft.Extensions.Configuration.Abstractions 2.2.0
Microsoft.Extensions.Configuration.Binder 2.2.0
Microsoft.Extensions.DependencyInjection.Abstractions 2.2.0
Microsoft.Extensions.Logging 2.2.0
Microsoft.Extensions.Logging.Abstractions 2.2.0
Microsoft.Extensions.Options 2.2.0
Microsoft.Extensions.Primitives 2.2.0
Microsoft.NETCore.Platforms 2.2.2
Microsoft.Owin 4.0.1
Microsoft.Rest.ClientRuntime 2.3.20
Microsoft.Rest.ClientRuntime.Azure 3.3.18
Microsoft.Win32.Primitives 4.3.0

## Description of the issue:
When I try to reject a c2d message, I get the exception below.

## Code sample exhibiting the issue:
```
c2dDeviceClient = Microsoft.Azure.Devices.Client.DeviceClient.Create(Environment.GetSetting("IoTHubHostName"), deviceAuth);
await c2dDeviceClient.RejectAsync(c2dMessage);
```

## Console log of the issue:
```
        Type: System.NullReferenceException
        Data: System.Collections.ListDictionaryInternal
        Message: Object reference not set to an instance of an object.
        Source: Microsoft.Azure.Devices.Client
        StackTrace:    at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpTransportHandler.<DisposeMessageAsync>d__25.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass22_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<ExecuteWithErrorHandlingAsync>d__23`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass28_0.<<RejectAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<RejectAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.InternalClient.<RejectAsync>d__63.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

at Azure_SharedUtilities.IoTHub.<RejectMessageAsync>d__8.MoveNext() in R:\...\Azure-SharedUtilities.cs:line 172
```

This issue was initially logged at https://social.msdn.microsoft.com/Forums/en-US/23e938ad-360d-4736-a5fc-e9f269eab8a2/getting-quotobject-reference-not-set-to-an-instance-of-an-objectquot-when-calling-rejectasync?forum=azureiothub

This might be related to https://github.com/Azure/azure-iot-sdk-csharp/issues/942 (and potentially solved by https://github.com/Azure/azure-iot-sdk-csharp/pull/1001 but I am unable to confirm)

**Comment by mfalkvidd** [comment]
I get the same error when I try using TransportType.Http1 when creating the device client. Maybe the HTTP client uses AMQP under the covers?

**Comment by bikamani** [comment]
@mfalkvidd Thanks for filing this issue. We will investigate and get back to you.

**Comment by bikamani** [comment]
@davilu Since you have worked on PR 1001 mentioned above, do you have more context to this issue?

**Comment by mfalkvidd** [comment]
I changed RejectAsync to CompleteAsync and I still get the same stack trace, so this problem seems to affect both methods.

**Comment by mfalkvidd** [comment]
I seem to have narrowed down the problem a bit more.
This is what my code does:
1. Create a DeviceClient for the device
2. Loop through all c2d messages for the device.
3. For each message, if the Message.MessageId is in the internal list of messages that have been processed, put the message object on a to-be-acknowledged-list.
4. If I get a timeout, I "know" there are no more messages. I close and dispose the DeviceClient.
5. Create a second DeviceClient for the device
6. For each message in the to-be-acknowledged-list, call CompleteAsync with the Message object.
7. Close and dispose the second DeviceClient.

If I skip steps 4+5, CompleteAsync is successful. So it seems like a Message object is only valid when it is used on the same DeviceClient that originally delivered the object?

**Comment by mfalkvidd** [comment]
Could that info be added to the documentation, and/or the stack trace be made to inform developers instead of just crashing and burning without a hint of what went wrong?

**Comment by davilu** [comment]
This issue was fixed in latest release [1.21.1](https://www.nuget.org/packages/Microsoft.Azure.Devices.Client/1.21.1) for AMQP.

**Comment by timtay-microsoft** [comment]
Closing this issue since the issue's fix was released in 1.21.1

**Comment by az-iot-builder-01** [comment]
@mfalkvidd, @davilu, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
